election_predictor/interface/main.py

```diff
@@ -0,0 +1,245 @@
+
+    #TODO Review NaNs logic with Niek and Chris – refactor into preproc?
+    X_train = X_train.fillna(value=0)
+
+    X_train = preprocessor_pipeline.fit_transform(X_train)
+
+    X_test = test_data.drop(columns=drop_columns)
+
+    X_test = X_test.fillna(value=0)
+
+    X_test = preprocessor_pipeline.transform(X_test)
+
+    # Convert train and test features back into DataFrames
+    X_train = pd.DataFrame(X_train, columns=preprocessor_pipeline.get_feature_names_out())
+    X_test = pd.DataFrame(X_test, columns=preprocessor_pipeline.get_feature_names_out())
+
+    # Handle target data
+    selected_targets = ['next_elec_date', 'LAB_ACT', 'CON_ACT', 'LIB_ACT', 'GRE_ACT',
+                        'BRX_ACT', 'SNP_ACT', 'UKI_ACT', 'PLC_ACT', 'OTH_ACT']
+
+    y_train = train_data[selected_targets]
+    y_test = test_data[selected_targets]
+
+    # Handle y_train results cleaning (create a fake election result)
+    # (removes actuals results, we are trying to predict)
+    y_train.loc[y_train['next_elec_date'] == election_date,
+         ['LAB_ACT', 'CON_ACT', 'LIB_ACT', 'GRE_ACT', 'BRX_ACT', 'SNP_ACT', 'UKI_ACT', 'PLC_ACT', 'OTH_ACT']] = np.nan
+
+    # Fetch mean values of each value in X_test
+    X_test_median = X_test.mean()
+
+    # Create imputation for y_train to impute over actuals we are trying to predict
+    imputation_values = {
+        'CON_ACT': X_test_median['CON_FC'],
+        'LAB_ACT': X_test_median['LAB_FC'],
+        'LIB_ACT': X_test_median['LIB_FC'],
+        'BRX_ACT': X_test_median['BRX_FC'],
+        'GRE_ACT': X_test_median['GRE_FC'],
+        'OTH_ACT': X_test_median['OTH_FC'],
+        'PLC_ACT': X_test_median['PLC_FC'],
+        'SNP_ACT': X_test_median['SNP_FC'],
+        'UKI_ACT': X_test_median['UKI_FC']
+    }
+
+    y_train = y_train.fillna(value=imputation_values)
+    y_train = y_train.fillna(value=0)
+
+    # Finally drop month_x columns from X_train and X_test
+    X_train.drop(columns=['Month_x'], inplace=True)
+    X_test.drop(columns=['Month_x'], inplace=True)
+
+
+    # Handle training and testing UK GE parties vote share
+
+    # Handle matrix conversion for train feature data
+    X_train_matrix = np.array(X_train)
+
+    parties = [
+        'LAB', 'CON', 'LIB', 'GRE', 'BRX', 'SNP', 'UKI', 'PLC',
+        'OTH'
+    ]
+
+    trained_models = { }
+
+    for party, party_code in enumerate(parties):
+        # Set party target train and test
+        party_y_train = y_train[party_code + "_ACT"]
+        party_y_test = y_test[party_code + "_ACT"]
+
+        # Handle XGBoost Regressor modelling
+        # Handle model fetch and initialisation
+        xgb_regressor = XGBoostModel()
+        xgbr_model = xgb_regressor.initialize_model()
+
+        # Handle model compiling
+        xgb_regressor.compile_model(xgbr_model) # Uses default model parameters from params.py
+
+        # Handle model training
+        trained_model = xgb_regressor.train_model(xgbr_model, X_train_matrix, party_y_train)
```

Comment:
does this instantiate a separate model for each party?

Reply:
Yes